database.sql

```diff
@@ -1287,7 +1287,8 @@ CREATE TABLE IF NOT EXISTS `user-item` (
 	`pinned` boolean COMMENT 'The item is pinned on the profile page',
 	`notification-type` tinyint unsigned NOT NULL DEFAULT 0 COMMENT '',
 	 PRIMARY KEY(`uid`,`iid`),
-	 INDEX `uid_pinned` (`uid`,`pinned`)
+	 INDEX `uid_pinned` (`uid`,`pinned`),
+	 INDEX `iid_uid` (`iid`,`uid`)
```

Comment:
No need, we already have the PRIMARY KEY for this couple of fields. Just invert the condition order on `uid` and `iid`.

Reply:
We cannot change the primary key. And we need the other key for the new notification class. There we will need to have a look for all notifications of a specific item id. And there we won't have a user id.

Reply:
Addition: We need the index order of the primary index for example for the API call. We really need both indexes.

Reply:
What API call?

Reply:
See this commit: https://github.com/friendica/friendica/pull/8059/commits/f134903fcfd2687a51c9c57e6597dd8508e00a18

Reply:
I see the use of the `(iid, uid)` index in the `INNER JOIN ... ON` but you could flip the `WHERE` condition to have ``AND`user-item`.`iid` > ? AND `user-item`.`uid` = ?`` and not need the `(uid, iid)` index, no?

Reply:
No :-) The index is needed for the sorting as well.

Reply:
When do you need to sort the notification **per user**?

Reply:
The best way (in the future "notification" implementation) will be to query the table `user-item` for the item id. Then the result will tell the users that have to be notified. When the index wouldn't be `(iid, uid)` in that case, the system had to perform a cost intensive table scan.

Reply:
So far you've described only a ``(`iid`, `uid`)`` index but not a ``(`uid`, `iid`)`` index (the current one). Can't we just rebuild the PRIMARY INDEX via the `update1329()` call?

Reply:
I described the need for both indexes, see here: https://github.com/friendica/friendica/pull/8059#discussion_r363121834

Not having the one or the other index will result in table scans.

Reply:
I understand the `(iid, uid)` index would be used in this kind of query:
```sql
SELECT *
FROM `item` i
JOIN `user-item` ui ON ui.`iid` = i.`id`
WHERE i.`id` = ?
```
Or
```sql
SELECT *
FROM `item` i
JOIN `user-item` ui ON ui.`iid` = i.`id` AND ui.`uid` = ?
ORDER BY ui.`iid`
```
But I don't see how a `(uid, iid)` would be used in what I understand would be
```sql
SELECT *
FROM `user-item` ui
JOIN `item` i ON i.`id` = ui.`iid`
WHERE ui.`uid` = ?
ORDER BY ui.`iid`
```
What is the difference with the second query I wrote?

Reply:
We need the sorting after the `iid` field. The first and second query are the worst possible way since - depending on the optimizer - this would result in a table scan through the `item` table or a table scan through the `user-item` table. Please trust me upon database issues. This is my daily business.

Reply:
I do trust you, but I also want to learn so that I don't just have to trust you. What I don't get is how the `(uid,iid)` index would be used for sorting just on `iid`.

Reply:
The query is on `uid` and the sorting is on `iid`. The query should always be on the first fields of the index, the sorting should always be on the last fields of the index. When the query isn't on the first fields of the index, then the SQL optimizer has to do a table scan. Same goes with the sorting.

Concerning your queries: The leading table should always be the one with the fewest entries and ideal is a small table since this reduces the memory amount.

Reply:
Thanks for the explanation!